src/Compiler/Service/ServiceAnalysis.fs

```diff
@@ -451,3 +452,1068 @@ module UnusedDeclarations =
+                | SynPat.Ands _, SynPat.As _
+                | SynPat.Tuple _, SynPat.As _
+
+                // x, (y, z)
+                | SynPat.Tuple _, SynPat.Tuple(isStruct = false)
+
+                // A, (B | C)
+                // A & (B | C)
+                | SynPat.Tuple _, SynPat.Or _
+                | SynPat.Ands _, SynPat.Or _
+
+                // (x : int) | x
+                // (x : int) & y
+                | SynPat.Or _, SynPat.Typed _
+                | SynPat.Ands _, SynPat.Typed _
+
+                // let () = …
+                // member _.M() = …
+                | SynPat.Paren _, SynPat.Const (SynConst.Unit, _)
+                | SynPat.LongIdent _, SynPat.Const (SynConst.Unit, _) -> ValueNone
+
+                | _, SynPat.Const _
+                | _, SynPat.Wild _
+                | _, SynPat.Named _
+                | _, SynPat.Typed _
+                | _, SynPat.LongIdent(argPats = SynArgPats.Pats [])
+                | _, SynPat.Tuple(isStruct = true)
+                | _, SynPat.Paren _
+                | _, SynPat.ArrayOrList _
+                | _, SynPat.Record _
+                | _, SynPat.Null _
+                | _, SynPat.OptionalVal _
+                | _, SynPat.IsInst _
+                | _, SynPat.QuoteExpr _
+
+                | SynPat.Or _, _
+                | SynPat.ListCons _, _
+                | SynPat.Ands _, _
+                | SynPat.As _, _
+                | SynPat.LongIdent _, _
+                | SynPat.Tuple _, _
+                | SynPat.Paren _, _
+                | SynPat.ArrayOrList _, _
+                | SynPat.Record _, _ -> ValueSome range
+
+                | _ -> ValueNone
+
+            | _ -> ValueNone
+
+    let getUnnecessaryParentheses (getSourceLineStr: int -> string) (parsedInput: ParsedInput) : Async<range seq> =
+        async {
+            let ranges = HashSet Range.comparer
+
+            let visitor =
+                { new SyntaxVisitorBase<obj>() with
+                    member _.VisitExpr(path, _, defaultTraverse, expr) =
+                        SynExpr.unnecessaryParentheses getSourceLineStr expr path
+                        |> ValueOption.iter (ranges.Add >> ignore)
+
+                        defaultTraverse expr
+
+                    member _.VisitPat(path, defaultTraverse, pat) =
+                        SynPat.unnecessaryParentheses pat path
+                        |> ValueOption.iter (ranges.Add >> ignore)
+
+                        defaultTraverse pat
+                }
+
+            // Traverse every node in the input.
+            let pick _ _ _ diveResults =
+                let rec loop =
+                    function
+                    | [] -> None
+                    | (_, project) :: rest ->
+                        ignore (project ())
+                        loop rest
+
+                loop diveResults
+
+            let _ = SyntaxTraversal.traverseUntil pick parsedInput.Range.End visitor parsedInput
```

Comment:
If this does what I think it does we should expose this publically in FCS.
My understanding is that this just keeps traversing and captures all the ranges in the HashSet?
@baronfel is this equivalent to what `SyntaxCollectorBase` does in FSAC?
This really might be worth it to expose as FCS API.

Reply:
yeah, fair point - but that would require someone to write custom `pick`s. I much prefer the experience of having the two distinct modes:

* traverse the tree down to this point
* traverse the entire tree (within a given range)

Reply:
> If this does what I think it does we should expose this publically in FCS. My understanding is that this just keeps traversing and captures all the ranges in the HashSet? @baronfel is this equivalent to what `SyntaxCollectorBase` does in FSAC? This really might be worth it to expose as FCS API.

Yes, the custom `pick` here (which never "picks" anything) lets it visit the whole AST. Some analyzers (where the goal is to generate diagnostics for a whole file) would indeed probably be much easier to implement if they had access to such a `traverseAll` function, which could be as simple as

```fsharp
let traverseAll (visitor: SyntaxVisitorBase<'T>) (parseTree: ParsedInput) : unit =
    let pick _ _ _ diveResults =
        let rec loop =
            function
            | [] -> None
            | (_, project) :: rest ->
                ignore (project ())
                loop rest

        loop diveResults

    ignore<'T option> (traverseUntil pick parseTree.Range.End visitor parseTree)
```

I can think of other possible signatures, though, e.g.,

```fsharp
val fold:
    folder: ('State -> SyntaxNode list -> SyntaxNode -> 'State) ->
    state: 'State
    tree: ParsedInput ->
        'State
```

where `folder` takes the state, path, and current node to produce the new state.

This can actually be implemented using the existing code like

```fsharp
let fold folder state tree =
    let mutable state = state

    let visitor =
        { new SyntaxVisitorBase<obj>() with
            member _.VisitExpr(path, _, defaultTraverse, expr) =
                state <- folder state path (SyntaxNode.SynExpr expr)
                defaultTraverse expr

            member _.VisitPat(path, defaultTraverse, pat) =
                state <- folder state path (SyntaxNode.SynPat pat)
                defaultTraverse pat

            // Override the rest likewise…
        }

    let pick _ _ _ diveResults =
        let rec loop =
            function
            | [] -> None
            | (_, project) :: rest ->
                ignore (project ())
                loop rest

        loop diveResults

    do ignore<obj option> (traverseUntil pick tree.Range.End visitor tree)
    state
```

It does make for pretty ergonomic usage, in my opinion:

```fsharp
let getUnnecessaryParentheses (getSourceLineStr: int -> string) (parsedInput: ParsedInput) : Async<range seq> =
    async {
        return upcast ((HashSet Range.comparer, parsedInput) ||> SyntaxTraversal.fold (fun acc path node ->
            match node with
            | SyntaxNode.SynExpr expr ->
                SynExpr.unnecessaryParentheses getSourceLineStr expr path
                |> ValueOption.map (fun range -> ignore (acc.Add range); acc)
                |> ValueOption.defaultValue acc
            | SyntaxNode.SynPat pat ->
                SynPat.unnecessaryParentheses pat path
                |> ValueOption.map (fun range -> ignore (acc.Add range); acc)
                |> ValueOption.defaultValue acc
            | _ -> acc))
    }
```

You could also have something like this:

```fsharp
val foldWhile:
    folder: ('State -> SyntaxNode list -> SyntaxNode -> (SyntaxNode * 'State) option) ->
    state: 'State
    tree: SyntaxNode ->
        'State
```

...where the folder returns which subnode to dive into next as well as the new state, which could also be used to implement the existing `Traverse`.

Reply:
This looks interesting. I think we should indeed not expose FCS consumers to write a custom `pick` function. But indeed exposing a `traverse until point` and `traverse entire tree` would be quite valuable.

I'm a little less convinced of the `folder: ('State -> SyntaxNode list -> SyntaxNode -> 'State)` idea. It looks nice but feels like a departure from the `SyntaxVisitorBase`. You would also ignore most `SyntaxNode` in a pattern match if you use this. While the `SyntaxVisitorBase` keeps it somewhat explicit what types of nodes you are hooking into.

We can circle back to this in a future PR.

Reply:
I ended up exposing

```fsharp
val internal traverseAll: visitor: SyntaxVisitorBase<'T> -> parseTree: ParsedInput -> unit
```

internally for now in 19da092 to make the call inside of `getUnnecessaryParentheses` clearer. `traverseUntil` is still there (`traverseAll` and `Traverse` both call it), but it's private to the `SyntaxTraversal` module.

Reply:
Resolving this since this can also be done in a followup.